File: backend/routes/employee_routes.py
from flask import Blueprint, request, jsonify
from config import EMPLOYEES_DB_FILE
from utils import load_json_data, save_json_data

# Using a relative import for utils and config assumes 'shavzak' is a package
# or that the app is run from the 'shavzak' directory.
# If running app.py directly from its location, you might need to adjust sys.path or use absolute imports if shavzak is installed.

employee_bp = Blueprint('employee_api', __name__, url_prefix='/api')

def _generate_employee_id(employees_data):
    """Generates a new unique employee ID."""
    if not employees_data:
        return "1"
    ids = [int(emp['id']) for emp in employees_data if emp.get('id', '').isdigit()]
    return str(max(ids) + 1 if ids else 1)

def _sort_employees(employees_list):
    """Sorts a list of employees by name, case-insensitively."""
    return sorted(employees_list, key=lambda emp: emp['name'].lower())

@employee_bp.route('/employees', methods=['GET'])
def get_employees():
    """Returns the list of all employees with their load data."""
    # The employees.json file is kept sorted by add/update/delete operations.
    employees = load_json_data(EMPLOYEES_DB_FILE, default_data=[])
    return jsonify(employees)

@employee_bp.route('/employees', methods=['POST'])
def add_employee():
    data = request.get_json()
    if not data or 'name' not in data or not data['name'].strip():
        return jsonify({"message": "Employee name is required"}), 400

    employees_list = load_json_data(EMPLOYEES_DB_FILE, default_data=[])
    new_name = data['name'].strip()
    if any(emp['name'].lower() == new_name.lower() for emp in employees_list):
        return jsonify({"message": f"Employee with name '{new_name}' already exists"}), 409

    new_employee_id = _generate_employee_id(employees_list)
    new_employee = {
        "id": new_employee_id, "name": new_name,
        "total_shifts_assigned": 0, "total_day_shifts_assigned": 0, "total_night_shifts_assigned": 0
    }
    employees_list.append(new_employee)
    save_json_data(_sort_employees(employees_list), EMPLOYEES_DB_FILE)
    return jsonify(new_employee), 201

@employee_bp.route('/employees/<string:employee_id>', methods=['PUT'])
def update_employee(employee_id):
    data = request.get_json()
    if not data or 'name' not in data or not data['name'].strip():
        return jsonify({"message": "Employee name is required for update"}), 400

    employees_list = load_json_data(EMPLOYEES_DB_FILE, default_data=[])
    employee_to_update = next((emp for emp in employees_list if emp['id'] == employee_id), None)

    if not employee_to_update:
        return jsonify({"message": "Employee not found"}), 404

    updated_name = data['name'].strip()
    if employee_to_update['name'].lower() != updated_name.lower() and \
       any(emp['name'].lower() == updated_name.lower() and emp['id'] != employee_id for emp in employees_list):
        return jsonify({"message": f"Another employee with name '{updated_name}' already exists"}), 409

    employee_to_update['name'] = updated_name
    save_json_data(_sort_employees(employees_list), EMPLOYEES_DB_FILE)
    return jsonify(employee_to_update), 200

@employee_bp.route('/employees/<string:employee_id>', methods=['DELETE'])
def delete_employee(employee_id):
    employees_list = load_json_data(EMPLOYEES_DB_FILE, default_data=[])
    original_length = len(employees_list)
    updated_employees_list = [emp for emp in employees_list if emp['id'] != employee_id]

    if len(updated_employees_list) == original_length:
        return jsonify({"message": "Employee not found"}), 404

    save_json_data(_sort_employees(updated_employees_list), EMPLOYEES_DB_FILE)
    return jsonify({"message": "Employee deleted successfully"}), 200

File: backend/src/routes/employeeRoutes.ts
import express, { Request, Response, Router } from 'express';
import { EMPLOYEES_COLLECTION } from '../config';
import {
  getCollectionDocs,
  getDocById,
  addDoc,
  updateDocById,
  deleteDocById,
} from '../utils/firestoreUtils';

// Define a type for your Employee data (similar to your frontend types.ts)
interface Employee {
  id?: string; // Firestore ID will be a string, optional when creating
  name: string;
  total_shifts_assigned: number;
  total_day_shifts_assigned: number;
  total_night_shifts_assigned: number;
}

const router: Router = express.Router();

const sortEmployees = (employeesList: Employee[]): Employee[] => {
  return [...employeesList].sort((a, b) => a.name.toLowerCase().localeCompare(b.name.toLowerCase()));
};

// GET /api/employees - Get all employees
router.get('/employees', async (req: Request, res: Response): Promise<void> => {
  try {
    const allEmployees = await getCollectionDocs<Employee>(EMPLOYEES_COLLECTION);
    res.status(200).json(sortEmployees(allEmployees));
  } catch (error) {
    console.error("Error fetching employees:", error);
    res.status(500).json({ message: "Error fetching employees" });
  }
});

// POST /api/employees - Add a new employee
router.post('/employees', async (req: Request, res: Response): Promise<any> => {
  try {
    const { name } = req.body;
    if (!name || typeof name !== 'string' || name.trim() === '') {
      return res.status(400).json({ message: "Employee name is required and must be a non-empty string" });
    }

    const allEmployees = await getCollectionDocs<Employee>(EMPLOYEES_COLLECTION);
    if (allEmployees.some(emp => emp.name.toLowerCase() === name.trim().toLowerCase())) {
      return res.status(409).json({ message: `Employee with name '${name.trim()}' already exists` });
    }

    const newEmployeeData: Omit<Employee, 'id'> = { // Omit 'id' as Firestore will generate it
      name: name.trim(),
      total_shifts_assigned: 0,
      total_day_shifts_assigned: 0,
      total_night_shifts_assigned: 0,
    };

    const newId = await addDoc(EMPLOYEES_COLLECTION, newEmployeeData);
    const addedEmployee: Employee = { ...newEmployeeData, id: newId };

    res.status(201).json(addedEmployee);
  } catch (error) {
    console.error("Error adding employee:", error);
    res.status(500).json({ message: "Error adding employee" });
  }
});

// PUT /api/employees/:employeeId - Update an employee
router.put('/employees/:employeeId', async (req: Request, res: Response): Promise<any> => {
  try {
    const { employeeId } = req.params;
    const { name } = req.body;

    if (!name || typeof name !== 'string' || name.trim() === '') {
      return res.status(400).json({ message: "Employee name is required for update" });
    }

    const employeeToUpdate = await getDocById<Employee>(EMPLOYEES_COLLECTION, employeeId);
    if (!employeeToUpdate) {
      return res.status(404).json({ message: "Employee not found" });
    }

    const updatedName = name.trim();
    // Check for duplicate names only if the name is actually changing
    if (employeeToUpdate.name.toLowerCase() !== updatedName.toLowerCase()) {
      const allEmployees = await getCollectionDocs<Employee>(EMPLOYEES_COLLECTION);
      if (allEmployees.some(emp => emp.name.toLowerCase() === updatedName.toLowerCase() && emp.id !== employeeId)) {
        return res.status(409).json({ message: `Another employee with name '${updatedName}' already exists` });
      }
    }

    await updateDocById(EMPLOYEES_COLLECTION, employeeId, { name: updatedName });
    const updatedEmployee: Employee = { ...employeeToUpdate, name: updatedName };

    res.status(200).json(updatedEmployee);
  } catch (error) {
    console.error("Error updating employee:", error);
    res.status(500).json({ message: "Error updating employee" });
  }
});

// DELETE /api/employees/:employeeId - Delete an employee
router.delete('/employees/:employeeId', async (req: Request, res: Response): Promise<any> => {
  try {
    const { employeeId } = req.params;

    const employeeToDelete = await getDocById<Employee>(EMPLOYEES_COLLECTION, employeeId);
    if (!employeeToDelete) {
      return res.status(404).json({ message: "Employee not found" });
    }

    await deleteDocById(EMPLOYEES_COLLECTION, employeeId);
    res.status(200).json({ message: "Employee deleted successfully" });
  } catch (error) {
    console.error("Error deleting employee:", error);
    res.status(500).json({ message: "Error deleting employee" });
  }
});

export default router;